File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) return head;
        // ListNode* p = reverseList(head->next);
        
        
        // head->next->next = head;
        // head->next = nullptr;
        // return p;
        ListNode* a = nullptr;
        ListNode* b = head;
        ListNode* c = head->next;
        while (b)
        {
            b->next = a;
            a = b;
            b = c;
            if (c)
                c = c->next;
        }
        return a;
    }
};

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) return head;

        ListNode newHead = null;
        ListNode next = head.next;

        while (head != null) {
            head.next = newHead;
            newHead = head;
            head = next;

            if (head != null) next = head.next;
        }

        return newHead;
    }
}